cmd/kubeadm/app/phases/kubelet/config.go

```diff
@@ -26,28 +26,54 @@ import (
 
 	"k8s.io/api/core/v1"
 	rbac "k8s.io/api/rbac/v1"
-	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/version"
 	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 	"k8s.io/kubernetes/cmd/kubeadm/app/componentconfigs"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
 	"k8s.io/kubernetes/cmd/kubeadm/app/util/apiclient"
 	rbachelper "k8s.io/kubernetes/pkg/apis/rbac/v1"
 	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	"k8s.io/kubernetes/pkg/util/procfs"
+	utilsexec "k8s.io/utils/exec"
 )
 
 // WriteConfigToDisk writes the kubelet config object down to a file
 // Used at "kubeadm init" and "kubeadm upgrade" time
-func WriteConfigToDisk(kubeletConfig *kubeletconfig.KubeletConfiguration, kubeletDir string) error {
+func WriteConfigToDisk(kubeletConfig *kubeletconfig.KubeletConfiguration, nodeReg *kubeadmapi.NodeRegistrationOptions, kubeletDir string) error {
 
-	kubeletBytes, err := getConfigBytes(kubeletConfig)
+	kubeletBytes, err := getConfigBytes(buildDynamicKubeletConfig(kubeletConfig, nodeReg))
 	if err != nil {
 		return err
 	}
 	return writeConfigBytesToDisk(kubeletBytes, kubeletDir)
 }
 
+// buildDynamicKubeletConfig returns a Kubelet config file with corrected values according to different OS configuration
+func buildDynamicKubeletConfig(kubeletConfig *kubeletconfig.KubeletConfiguration, nodeReg *kubeadmapi.NodeRegistrationOptions) *kubeletconfig.KubeletConfiguration {
+
+	if nodeReg.CRISocket == kubeadmconstants.DefaultDockerCRISocket {
+		driver, err := kubeadmutil.GetCgroupDriverDocker(utilsexec.New())
+		if err != nil {
+			klog.Warningf("cannot automatically assign a cgroupDriver value in the KubeletConfiguration file: %v\n", err)
+		} else {
+			kubeletConfig.CgroupDriver = driver
+		}
+	}
+
+	// TODO: Conditionally set `CgroupDriver` to either `systemd` or `cgroupfs` for CRI other than Docker
+
+	if pids, _ := procfs.PidOf("systemd-resolved"); len(pids) > 0 {
```

Comment:
the function should be a no-op for Windows.

i suggest you create:
```
cmd/kubeadm/app/phases/kubelet/config.go
cmd/kubeadm/app/phases/kubelet/config_windows.go
cmd/kubeadm/app/phases/kubelet/config_unix.go
```

and have this function be different between the OS-es.

cc @ksubrmnn